kairos-server/src/config.rs

```diff
@@ -1,26 +1,119 @@
-use std::{fmt, str::FromStr};
+use config::{Config as Configuration, Environment, File};
+use dotenvy::dotenv;
+use serde::{Deserialize, Serialize};
+use std::{fs::File as FsFile, net::SocketAddr, path::Path};
+use tracing::{subscriber::set_global_default, Level};
+use tracing_subscriber::{fmt, prelude::*, EnvFilter};
 
-#[derive(Debug)]
-pub struct ServerConfig {
-    pub port: u16,
-}
+fn serialize_level_filter<S>(level: &Level, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    let level_str = match *level {
+        Level::ERROR => "error",
+        Level::WARN => "warn",
+        Level::INFO => "info",
+        Level::DEBUG => "debug",
+        Level::TRACE => "trace",
+    };
 
-impl ServerConfig {
-    pub fn from_env() -> Result<Self, String> {
-        let port = parse_env_as::<u16>("KAIROS_SERVER_PORT")?;
-        Ok(Self { port })
-    }
+    serializer.serialize_str(level_str)
 }
 
-fn parse_env_as<T>(env: &str) -> Result<T, String>
+fn deserialize_level_filter<'de, D>(deserializer: D) -> Result<Level, D::Error>
 where
-    T: FromStr,
-    <T as FromStr>::Err: fmt::Display,
+    D: serde::Deserializer<'de>,
 {
-    std::env::var(env)
-        .map_err(|e| format!("Failed to fetch environment variable {}: {}", env, e))
-        .and_then(|val| {
-            val.parse::<T>()
-                .map_err(|e| format!("Failed to parse {}: {}", env, e))
-        })
+    let s = String::deserialize(deserializer)?;
+    match s.to_lowercase().as_str() {
+        "error" => Ok(Level::ERROR),
+        "warn" => Ok(Level::WARN),
+        "info" => Ok(Level::INFO),
+        "debug" => Ok(Level::DEBUG),
+        "trace" => Ok(Level::TRACE),
+        _ => Err(serde::de::Error::custom("unknown level")),
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Server {
+    pub address: String,
+    pub port: u16,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Log {
+    #[serde(
+        deserialize_with = "deserialize_level_filter",
+        serialize_with = "serialize_level_filter"
+    )]
+    pub level: Level,
+    pub file: Option<String>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Settings {
+    pub server: Server,
+    pub log: Log,
+}
+
+#[allow(clippy::new_without_default)]
+impl Settings {
```

Comment:
Isn't there a crate that takes a type and just derives the toml serializer/ parser

Reply:
This allow clippy statement is because Avi told me not to construct state. There were default parameters in place before and the allow clippy wasn't needed. If I put the default function back, then the config crate will first construct the default, then load from toml, then dotenv, and then environment. ATM it will simply panic if there are parameters it can't gather from all those sources.